File: solutions/javascript/armstrong-numbers/1/armstrong-numbers.js
// my try to NOT convert to string
function getDigits(number) {
  let digits = [];
  while (number >= 1) {
    const lastDigit = number % 10;
    number = (number / 10) | 0;
    digits.unshift(lastDigit);
  }
  return digits;
}

export const isArmstrongNumber = number => {
  const digits = getDigits(number);
  const reducer = (acc, val) => acc + Math.pow(val, digits.length);
  const result = digits.reduce(reducer, 0);
  return result === number;
};

File: solutions/go/armstrong-numbers/1/armstrong_numbers.go
package armstrong

func IsNumber(n int) bool {
	if n < 10 {
		return true
	}

	if n < 100 {
		return false
	}

	exp := length(n)
	remnant := n
	sum := 0

	for remnant > 0 {
		sum += pow(remnant%10, exp)
		remnant /= 10
	}

	return sum == n
}

// pow returns x^y faster than math.Pow(x, y) because it avoids casting to float64 and back to int.
func pow(x, y int) int {
	if y == 0 {
		return 1
	}
	result := x
	for i := 2; i <= y; i++ {
		result *= x
	}
	return result
}

// length returns the count of digits in the int.
func length(n int) int {
	count := 0
	for n > 0 {
		count++
		n /= 10
	}
	return count
}
